Allow named Arguments to be passed to Dto

eltharin opened the pull request:
Allow to change argument order or use variadic argument in dto constructor.

When Dto implements (empty) interface WithNamedArguments, ObjectHydrator will pass an associated key<>value argument to Dto constructor.
Dto can use named arguments or a variadic argument to put values in their properties.

Final Goal (to do after) is to allow `select NEW ItemDto(item.*) FROM App\Entity\Item as item`


updated with with the nested New operator PR.

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
 	
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple query but not ever the same parameters, you can use named arguments or variadic arguments, add ``WithNamedArguments`` to your Dto :
```
```suggestion
If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
you can use named arguments or variadic arguments, add ``WithNamedArguments`` to your Dto :
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
+
+    class CustomerDTO implements WithNamedArguments
+    {
+        public function __construct(public string|null $name = null, public string|null $email = null, public string|null $city = null, public mixed|null $value = null)
```
Let's use linebreaks.

```suggestion
        public function __construct(
                public string|null $name = null,
                public string|null $email = null,
                public string|null $city = null,
                public mixed|null $value = null,
        )
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
+        }
+    }
+
+And then you can select the fields you want in order you want :
```
```suggestion
And then you can select the fields you want in the order you want :
```

greg0ire commented on phpcs.xml.dist:
```diff
@@ -273,4 +273,9 @@
+
+     <rule ref="Generic.Files.LineEndings.InvalidEOLChar">
+      <exclude name="Generic.Files.LineEndings.InvalidEOLChar"/>
+     </rule>
```
Why is this necessary?

eltharin replied:
my bad, I'm on windows and when I make local test it is

greg0ire commented on src/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -254,14 +256,15 @@ abstract protected function hydrateAllData(): mixed;
      *                   newObjects?: array<array-key, array{
      *                       class: ReflectionClass,
      *                       args: array,
+     *                       argNames?: array,
```
Can you do better than just `array` here?

greg0ire commented on src/Internal/Hydration/AbstractHydrator.php:
```diff
@@ -341,33 +342,54 @@ protected function gatherRowData(array $data, array &$id, array &$nonemptyCompon
-            $obj   = $class->newInstanceArgs($args);
+            $class          = $newObject['class'];
+            $args           = $newObject['args'];
+            $argColumnNames = $newObject['argColumnNames'];
```
I would avoid creating 3 new variables, here and below, but maybe I'm missing something?

greg0ire commented on src/Query/ResultSetMapping.php:
```diff
@@ -552,25 +551,4 @@ public function addMetaResult(
-        );
-
-        return $this;
-    }
```
This looks a lot like a breaking change.

eltharin replied:
This function was created in my previous PR, so not very break

derrabus replied:
Indeed, we haven't issued a release yet that includes this method. The change is fine.

greg0ire commented on tests/Tests/Models/CMS/CmsAddressDTONamedArgs.php:
```diff
@@ -0,0 +1,14 @@
+
+class CmsAddressDTONamedArgs implements WithNamedArguments
+{
+    public function __construct(public string|null $country = null, public string|null $city = null, public string|null $zip = null, public CmsAddressDTO|string|null $address = null)
```
linebreaks please

greg0ire commented on tests/Tests/Models/CMS/CmsUserDTONamedArgs.php:
```diff
@@ -0,0 +1,14 @@
+
+class CmsUserDTONamedArgs implements WithNamedArguments
+{
+    public function __construct(public string|null $name = null, public string|null $email = null, public string|null $address = null, public int|null $phonenumbers = null, public CmsAddressDTO|null $addressDto = null, public CmsAddressDTONamedArgs|null $addressDtoNamedArgs = null)
```
Here as well.

greg0ire commented on tests/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
+                new CmsUserDTONamedArgs(
+                    e.email,
+                    u.name,
+                    CONCAT(a.country, \' \', a.city, \' \', a.zip) AS address
```
Please use [nowdoc](https://www.php.net/manual/en/language.types.string.php#language.types.string.syntax.nowdoc) so as to avoid this escaping.

greg0ire commented on tests/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
+        $dql = '
+            SELECT
+                new CmsUserDTOVariadicArg(
+                    CONCAT(a.country, \' \', a.city, \' \', a.zip) AS address,
```
Here too please

greg0ire commented on tests/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
+
+        self::assertSame($this->fixtures[0]->address->country . ' ' . $this->fixtures[0]->address->city . ' ' . $this->fixtures[0]->address->zip, $result[0]['user']->address);
+        self::assertSame($this->fixtures[1]->address->country . ' ' . $this->fixtures[1]->address->city . ' ' . $this->fixtures[1]->address->zip, $result[1]['user']->address);
+        self::assertSame($this->fixtures[2]->address->country . ' ' . $this->fixtures[2]->address->city . ' ' . $this->fixtures[2]->address->zip, $result[2]['user']->address);
```
Linebreaks please, and please use `sprintf()`

greg0ire commented on tests/Tests/ORM/Functional/NewOperatorTest.php:
```diff
@@ -1080,6 +1083,242 @@ public function testShouldSupportNestedNewOperators(): void
+
+        self::assertSame($this->fixtures[0]->address->country . ' ' . $this->fixtures[0]->address->city . ' ' . $this->fixtures[0]->address->zip, $result[0]['user']->address);
+        self::assertSame($this->fixtures[1]->address->country . ' ' . $this->fixtures[1]->address->city . ' ' . $this->fixtures[1]->address->zip, $result[1]['user']->address);
+        self::assertSame($this->fixtures[2]->address->country . ' ' . $this->fixtures[2]->address->city . ' ' . $this->fixtures[2]->address->zip, $result[2]['user']->address);
```
Here as well please.

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a');
```
Wait so what happens here? Which fields of the DTO are supposed to get hydrated? Hydrating `CustomerDTO::$name` with a city feels wrong :thinking:

eltharin replied:
No, it's exactly the case I want, As the args are named, it's city and name from the Dto witch are filled .
Maybe I have to print the Dto Output for better understand

greg0ire replied:
Yes, and also, you could explain what is going to happen: `And then you can select the fields you want in the order you want, and the ORM will try to match argument names with the selected field names`. If you do that, you should probably explain what happens if one of the selected field matches, but not the other. Or what happens if we do `SELECT NEW CustomerDTO(a.city, b.city) FROM …`

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
+    $query = $em->createQuery('SELECT NEW CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+you can either aliases column :
```
What do you mean here? I would expect an "or" close in your sentence, like "You can either do this or do this".

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,38 @@ You can also nest several DTO :
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a');
```
Did you consider a syntax closer to PHP, like `value: CONCAT(a.city, ' ', a.zip)`?

eltharin replied:
I can think about it but since we are in a SQL/DQL query I stayed in this way

eltharin replied:
It works, but I really hate this notation in an SQL query, so I left AS keywork working too.

derrabus commented on src/WithNamedArguments.php:
```diff
@@ -0,0 +1,9 @@
+
+namespace Doctrine\ORM;
+
+interface WithNamedArguments
```
Can we solve this issue without introducing empty marker interfaces?

eltharin replied:
We can have `$em->createQuery('SELECT NEW NAMED CustomerDTO(...);`

But I find it less readable

eltharin replied:
Do you prefer have a const value USE_NAMED_ARGS in dto instead of implents empty interface  ?

derrabus replied:
The gods of PHP gave us attributes, so we don't need marker interfaces anymore. If you really really need to flag the DTO class in some way, please use an attribute for that.

That being said, I think I'd prefer a dedicated DQL syntax for named argument calls. That should make any marking of DTO classes obsolete.

derrabus commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,60 @@ You can also nest several DTO :
+
+    use Doctrine\ORM\WithNamedArguments;
+
+    class CustomerDTO implements WithNamedArguments
```
Now that you've removed that interface, the documentation needs to be adjusted.

derrabus commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,60 @@ You can also nest several DTO :
+                public string|null $email = null,
+                public string|null $city = null,
+                public mixed|null $value = null,
+        )
+        {
```
```suggestion
        ) {
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a');
+    $users = $query->getResult(); // array of CustomerDTO
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
```
You used a tab here. Please check the whole diff and make sure you never use tabs.

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+The `NAMED` keywork must be added to all dto you want to instantiate :
```
```suggestion
The ``NAMED`` keyword must be added to all dto you want to instantiate :
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
+
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+Be careful, if two fields have the same name/alias, second will erase first.
```
```suggestion
Be careful, if two fields have the same name/alias, the second one will overwrite the first one.
```

greg0ire replied:
Should it instead throw an exception?

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+Be careful, if two fields have the same name/alias, second will erase first.
+If you put an argument name empty (for functions) or not in constructor argument list, ORM will throw an error.
```
> If you put an argument name empty (for functions)

What do you mean by that?

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -608,6 +608,68 @@ You can also nest several DTO :
+	// CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+Be careful, if two fields have the same name/alias, second will erase first.
+If you put an argument name empty (for functions) or not in constructor argument list, ORM will throw an error.
```
An error or an exception?

eltharin replied:
actually an error, thorwn by PHP and ORM : "Error: Unknown named parameter $"

greg0ire replied:
Sorry, where in the code is it thrown? I can't find it.

greg0ire commented:
Please kindly squash your commits together. If you don't, we'll try to remember to do it for you but it's best if you save us this trouble.

How to do that?

1. `git rebase -i origin/3.3.x`, assuming `origin` is a git remote that points to this repository, and not your fork. If you're not sure what your remotes are, run `git remote -vvv`, there should be your fork and the holy/reference/base/origin/whatever-you-call-it repository.
2. A window will show up with many lines, replace `pick` with `fixup` on every line but the first one
3. Close your editor, git should do its magic, and you should end up with one commit
4. Use `git push --force` to overwrite what you already push. Don't forget the `--force` option otherwise git will try to merge both things together.

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+If two fields have the same name/alias, a RepeatedAliasException is thrown.
```
```suggestion
If two fields have the same name/alias, a ``RepeatedAliasException`` is thrown.
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+If two fields have the same name/alias, a RepeatedAliasException is thrown.
+If a field without column name (functions for example) has no alias set, a MissingAliasException is thrown.
```
```suggestion
If a field without a column name (when using functions for example) has no alias set, a ``MissingAliasException`` is thrown.
```

greg0ire replied:
[When I use a function](https://sqlfiddle.com/postgresql/online-compiler?id=26728015-ebc0-4603-8e34-69c6cc60fa4a) the resulting column does have a name :thinking:

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+For aliases a field you can use SQL notation with `AS` keyword, or the PHP named arguments notation :
```
```suggestion
In order to  alias a field you can either use the SQL keyword ``AS``, or PHP's named arguments syntax :
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a');
```
The example only illustrates the second method. Also, you are saying "For aliases a field", but with the second method, there is no aliasing taking place at all, right?

eltharin replied:
It's for showing we can make a mix of field names and aliases

greg0ire replied:
In the example though, it's 100% field names, right? Or is there an alias I am missing?

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+
+    // CustomerDTO => {name : 'DOE', email: null, city: null, value: 'New York 10011'}
+
+The ``NAMED`` keyword must be added to all dto you want to instantiate :
```
```suggestion
The ``NAMED`` keyword must precede all DTO you want to instantiate :
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add ``named`` keyword in your DQL querie before your Dto :
```
```suggestion
you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+                public string|null $email = null,
+                public string|null $city = null,
+                public mixed|null $value = null,
+                public AddressDTO|null $address = null,
```
```suggestion
            public string|null $name = null,
            public string|null $email = null,
            public string|null $city = null,
            public mixed|null $value = null,
            public AddressDTO|null $address = null,
```

greg0ire commented on src/Exception/MissingAliasException.php:
```diff
@@ -0,0 +1,15 @@
+{
+    public static function create(): self
+    {
+        return new self('Alias is missing on a select part.');
```
Would it be possible to get more details here? Like which select part is missing an alias?

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
```
Here you mentioned named arguments and variadic arguments. Is it necessary to add the `NAMED` keyword to use variadic arguments?

eltharin replied:
variadic argument can be used without ``named`` keyword but the array will be "indexed" while with ``named``it will be associated

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+
 Note that you can only pass scalar expressions or other Data Transfer Objects to the constructor.
 
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
```
```suggestion
If you use your data transfer objects for multiple queries but not necessarily with the same parameters,
```

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+If you use your data-transfer objects for multiple queries but not necessarily with the same parameters,
+you can use named arguments or variadic arguments, add the ``named`` keyword in your DQL query before your DTO :
+
+.. code-block:: php
```
Because of what precedes, I'm expecting to see a DQL query. I think you should reorganize this.
If you use your data transfer objects for multiple queries, and you would rather not have to
specify arguments that precede the ones you are really interested in, you can use named arguments.

Consider the following DTO, which uses optional arguments:

// code for CustomerDTO here

You can specify arbitrary arguments in an arbitrary order by using the named argument syntax, and the ORM will try to match argument names with the selected column names.
The syntax relies on the `NAMED` keyword, like so:

greg0ire commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,85 @@ You can also nest several DTO :
+
+    // CustomerDTO => {name : 'SMITH', email: null, city: 'London', value: null}
+
+ORM will also look column aliases before columns names :
```
```suggestion
ORM will also give precedence to column aliases over column names :
```

greg0ire commented on src/Exception/DuplicateFieldException.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public static function create(string $argName, string $columnName): self
+    {
+        return new self(sprintf('Name %s for `%s` already in use.', $argName, $columnName));
```
```suggestion
        return new self(sprintf('Name "%s" for "%s" already in use.', $argName, $columnName));
```

greg0ire commented on src/Exception/NoMatchingPropertyException.php:
```diff
@@ -0,0 +1,17 @@
+{
+    public static function create(string $property): self
+    {
+        return new self(sprintf('Column Name `%s` has no property name or alias.', $property));
```
```suggestion
        return new self(sprintf('Column name "%s" does not match any property name. Consider aliasing it to the name of an existing property.', $property));
```

greg0ire commented on src/Query/Parser.php:
```diff
@@ -1764,29 +1774,71 @@ public function NewObjectExpression(): AST\NewObjectExpression
+            $key = $fieldAlias ?? $newArg->field ?? null;
+
+            if ($key === null) {
+                throw NoMatchingPropertyException::create(trim(substr(($this->query->getDQL() ?? ''), $startToken, ($this->lexer->lookahead->position ?? 0) - $startToken)));
```
Let's use linebreaks please.

```suggestion
                throw NoMatchingPropertyException::create(trim(substr(
                    ($this->query->getDQL() ?? ''),
                    $startToken,
                    ($this->lexer->lookahead->position ?? 0) - $startToken,
                )));
```

greg0ire commented on src/Query/Parser.php:
```diff
@@ -1764,29 +1774,71 @@ public function NewObjectExpression(): AST\NewObjectExpression
+            }
+
+            if (array_key_exists($key, $args)) {
+                throw DuplicateFieldException::create($key, trim(substr(($this->query->getDQL() ?? ''), $startToken, ($this->lexer->lookahead->position ?? 0) - $startToken)));
```
```suggestion
                throw DuplicateFieldException::create($key, trim(substr(
                    ($this->query->getDQL() ?? ''),
                    $startToken,
                    ($this->lexer->lookahead->position ?? 0) - $startToken
                )));
```

greg0ire commented:
@derrabus please give this another review.

eltharin commented:
wainting for https://github.com/doctrine/orm/pull/11619 to push EBNF update

beberlei commented on docs/en/reference/dql-doctrine-query-language.rst:
```diff
@@ -591,23 +591,88 @@ You can also nest several DTO :
+.. code-block:: php
+
+    <?php
+    $query = $em->createQuery('SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a');
```
I would prefer if we have only this syntax available `CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip))` and not the two above that implicitly map the name or the AS naming.

This would in my opinion also make the `NAMED` keyword obsolete, which I would like to get rid of as well.

The goal should be that named arguments syntax looks exactly the same as in PHP itself, because the `NEW className` somewhat implies that this is what DQL supports.

eltharin replied:
Disable implicity map means that we have to write all the field's names for our queries, instead of add a keyword "named".

Remove AS keyword means for those who actually have a named array as result (with AS keyword) to rewrite all the names while in my case just have to add NEW NAMED myDTO(...) and it's done.

Forcing to write: name: c.name, lastname: c.lastname, age: c.age, birthdate: c.birthdate, ... is in my opinion a waste of time for the few fields to rename.

For those as me who have helpers to select fields to add in queries, remove NAMED keyword means to be obliged to name ALL the fields.

As I said in descritpion, final goal is to be able to write c.* and have all the fields by the names, 

That's why the first version of this PR have only AS keyword and not the 'PHP syntax'

beberlei replied:
@eltharin I discussed with @greg0ire and @derrabus at our Doctrine core team meetup in person, and I guess the main thing is that we want to avoid adding 2 syntaxes (implicit and AS naming, plus name double colon). We considered both and your arguments and then we would rather remove the PHP-style named arguments syntax and only keep the examples 1+2, so the following should be allowed:

* SELECT NEW NAMED CustomerDTO(a.city, c.name) FROM Customer c JOIN c.address a')
* SELECT NEW NAMED CustomerDTO(c.name, CONCAT(a.city, ' ' , a.zip) AS value) FROM Customer c JOIN c.address a

and this syntax should be removed again:

* SELECT NEW NAMED CustomerDTO(name: c.name, value: CONCAT(a.city, ' ' , a.zip)) FROM Customer c JOIN c.address a

Everything else is 👍 from me.

eltharin replied:
thank you to all three of you  (and other mainteners) for being so open-minded and for your work.

eltharin committed:
Allow named Arguments to be passed to Dto

Allow to change argument order or use variadic argument in dto constructor using new named keyword

greg0ire commented:
Thanks @eltharin !

beberlei commented:
@eltharin 💯 thanks
